src/Parallel-Solvers/Simulation_Parameters/Geometry.h

```diff
@@ -19,35 +31,137 @@ struct Volume : Yaml::ValidatedYaml {
-    bool contains(const double* elem_coords) { 
+    bool contains(const double* elem_coords) {
       double radius;
 
       switch(type) {
-      case VOLUME_TYPE::global:
-        return true;
+          case VOLUME_TYPE::global:
+            return true;
 
-      case VOLUME_TYPE::box:
-        return ( elem_coords[0] >= x1 && elem_coords[0] <= x2
-              && elem_coords[1] >= y1 && elem_coords[1] <= y2
-              && elem_coords[2] >= z1 && elem_coords[2] <= z2 );
+          case VOLUME_TYPE::box:
+            return ( elem_coords[0] >= x1 && elem_coords[0] <= x2
+                  && elem_coords[1] >= y1 && elem_coords[1] <= y2
+                  && elem_coords[2] >= z1 && elem_coords[2] <= z2 );
 
-      case VOLUME_TYPE::cylinder:
-        radius = sqrt( elem_coords[0]*elem_coords[0] +
-                        elem_coords[1]*elem_coords[1] ); 
-        return ( radius >= radius1
-              && radius <= radius2 );
+          case VOLUME_TYPE::cylinder:
+            radius = sqrt( elem_coords[0]*elem_coords[0] +
+                            elem_coords[1]*elem_coords[1] );
+            return ( radius >= radius1
+                  && radius <= radius2 );
 
-      case VOLUME_TYPE::sphere:
-        radius = sqrt( elem_coords[0]*elem_coords[0] +
-                        elem_coords[1]*elem_coords[1] +
-                        elem_coords[2]*elem_coords[2] );
-        return ( radius >= radius1
-              && radius <= radius2 );
+          case VOLUME_TYPE::sphere:
+            radius = sqrt( (elem_coords[0] - orig_x)*(elem_coords[0] - orig_x) +
+                             (elem_coords[1] - orig_y)*(elem_coords[1] - orig_y) +
+                             (elem_coords[2] - orig_z)*(elem_coords[2] - orig_z) );
+            return ( radius >= radius1
+                  && radius <= radius2 );
+                  
+          case VOLUME_TYPE::stl:
+            fill_this = false;  // default is no, don't fill it
+            
+            // find the closest element in the voxel mesh to this element
+            i0_real = (elem_coords[0] - orig_x)/(voxel_dx);
+            j0_real = (elem_coords[1] - orig_y)/(voxel_dy);
+            k0_real = (elem_coords[2] - orig_z)/(voxel_dz);
+            
+            i0 = (int)i0_real;
+            j0 = (int)j0_real;
+            k0 = (int)k0_real;
+            
+            // look for the closest element in the voxel mesh
+            elem_id0 = get_id(i0,j0,k0,num_voxel_x,num_voxel_y);
+            
+            // if voxel mesh overlaps this mesh, then fill it if =1
+            if (elem_id0 < num_voxel_x*num_voxel_y*num_voxel_z && elem_id0>=0){
+                
+                // voxel mesh elem values = 0 or 1
+                if (voxel_elem_values(elem_id0) == true) {
+                    fill_this = true;
+                } else {
+                    fill_this = false;
+                }
+                
+            } // end if
+            
+            // check for periodic
```

Comment:
Can you expand a bit on this comment.  Im not sure what it is doing.

Reply:
This section of code checks for periodic repetition of the imported region and prevents it from being repeated when it is being painted onto the global mesh. These details were added to the code for clarity.